tools/hilow.wdl

```diff
@@ -0,0 +1,452 @@
+
+    command <<<
+        prom=~{basename(promoters,".bed")}
+        sort \
+            -k1,1 \
+            -k2,2n \
+            -k4,4 \
+            ~{promoters} \
+            | awk \
+            -F'\t' \
+            -v OFS="\t" \
+            '{ a[$1"\t"$2"\t"$3]=($1"\t"$2"\t"$3 in a? a[$1"\t"$2"\t"$3]",":"")$4 }END{ for(i in a) print i,a[i] }'
+            | sort \
+            -k1,1 \
+            -k2,2n \
+            > $prom".unique.bed"
+
+        #get read counts
+        intersectBed -c -a $prom".unique.bed" -b ~{bam} > $prom"_Signal"
+
+        #compute top 2/3
+        length=$(wc -l $prom"_Signal" | awk -F\  '{print $1}'); 
+        keepN=$(echo "$(($length * 2 / 3))"); 
+
+        sort -k5,5nr $prom"_Signal" \
+            | head -n $keepN \
+            | sort -k1,1 -k2,2n \
+            | cut -f 1-3 \
+            > $prom"_Signal.keep."$keepN".3cols.bed"
+
+        #get non-promoter region
+        cut -f 1-3 ~{loop_bed} 
+            | sort -k1,1 -k2,2n \
+            > ~{basename(loop_bed)}
+        bedtools \
+            subtract \
+            -a ~{basename(loop_bed)} \
+            -b $prom".unique.bed" \
+            | sort \
+            > Enhancers.Subtract.4kbPromoters.sort.bed
+
+        cat \
+            $prom"_Signal.keep."$keepN".3cols.bed" \
+            Enhancers.Subtract.4kbPromoters.sort.bed \
+            | sort -k1,1 -k2,2n \
+            > ~{outfile}
+    >>>
+
+    output {
+        File combined_promoters = outfile
+    }
+
+    runtime {
+        cpu: 8
+        memory: "20 GB"
+        maxRetries: 1
+        container: "ghcr.io/stjudecloud/bedtools:2.31.1"
+    }
+}
+
+task extract_promoters {
+    meta {
+        description: "Extract promoters from a GTF file"
+        outputs: {
+            promoter: "Promoter regions in BED format"
+        }
+    }
+
+       parameter_meta {
+        annotation: "GTF (optionally gzip compressed) file containing gene annotations"
+    }
+
+    input {
+        File annotation
+    }
+
+    String base = basename(annotation, ".gz")
+
+    command <<<
+        set -euo pipefail
+
+        gunzip -c ~{annotation} > ~{base} \
+           || ln -sf ~{annotation} ~{base}
+
+
+        sed -i 's/ /\t/g' ~{base}
+
+        # Find all transcripts, excluding the mitochondrial chromosome
+        # Add a 2000bp buffer to the transcription start site to include promoter
+        # Extract gene_id and transcript_id ($10 and $12).
+        # Subtract the suffix from the trasncript ID.
+        # Convert to BED format
+        awk -F\\t '{ if($3 == "transcript" && $1 !~ "chrM")
+            if($7 =="+")
+                print $1 "++" $4-2000 "++" $4+2000 "++" substr($12,2,length($12)-5) "|" substr($10,2,length($10)-3)
+            else if($7=="-")
+                print $1 "++" $5-2000 "++" $5+2000 "++" substr($12,2,length($12)-5) "|" substr($10,2,length($10)-3) }' \
+            ~{base} \
+            | grep -v "unassigned" \
+            | sort \
+            | uniq \
```

Comment:
```suggestion
            | sort -u \
```

Reply:
I'm not sure I've ever actually used `sort -u`. I always pipe through `uniq`.